### ISSUE TITLE
[TODO] Implement spatial split BVH builder

### ISSUE THREAD

**Issue by syoyo** [lighttransport/nanort#15]

**Comment by syoyo** [comment]
Partial implementation is done in `sbvh` branch : https://github.com/lighttransport/nanort/tree/sbvh

**Comment by tay10r** [comment]
Would an LBVH algorithm count as a spatial split BVH? I have an existing LBVH header only library that builds Sponza in parallel in about 25 ms, under the MIT license. You could use that, if you'd like.

**Comment by syoyo** [comment]
> Would an LBVH algorithm count as a spatial split BVH?

No. LBVH does not consider spatial splitting. I have also implemented LBVH in another project(and it also does not consider spatial splitting)
https://github.com/avr-aics-riken/SURFACE/blob/master/render/render_prefix_tree_util.h

As you may know, spatial splitting requires PrimRef data structure(indirect access to primitive).
 
> There's also this header only BVH library, which provides high quality BVHs with SAH and builds them in parallel as well

this BVH library looks nice, but it also does not implement spatial split BVH.

**Comment by tay10r** [comment]
Hey! I'm terrible sorry! It's probably obvious at this point that I didn't know exactly what spatial splitting was when I wrote that comment. My bad!

**Comment by cesss** [comment]
> this BVH library looks nice, but it also does not implement spatial split BVH.

It seems they did implement SBVH now! (actually, the commit seems to be from April 17). However, if you look at the performance comparison chart, you'll see that SBVH is not substantially faster than other algorithms they provide...

**Comment by syoyo** [comment]
@cess Oh nice! > It seems they did implement SBVH now!

**Comment by cesss** [comment]
> @cess Oh nice! > It seems they did implement SBVH now!

Yes, but I encourage you to improve nanort anyway, because that other BVH library is very demanding in terms of C++  version (it requires C++17, and I prefer to have the freedom of not being tied to too new C++ versions  --in fact I decided to move from C++ to C for my main work, because I like C more... I'm wrapping the BVH library for the moment, with a C layer on top of it)

**Comment by syoyo** [comment]
@cess Yes, we plan to cleanup NanoRT code but it still stick with C++11(e.g. more modern use of C++11 templates, better Python integration with pybind11). After of it we may consider backport  `madmann91/bvh` to nanort(still NanoRT uses C++11). Fortunately the code of `madmann91/bvh` is rather simple and easy to tweak.
